### PR TITLE
Update Cupertino desktop text selection toolbar

### PR DESCRIPTION
**Pull request by luccasclezar**
The current look-and-feel of the CupertinoDesktopTextSelectionToolbar is outdated and it doesn't change the text color when hovered, which usually leads to poor contrast (see first image).

Current Flutter context menu (pre-PR):
Screenshot 2023-03-02 at 17 37 49

New Flutter context menu:
Screenshot 2023-03-02 at 17 38 15

Native context menu on macOS 13.2 for reference:
Screenshot 2023-03-02 at 17 44 06

## Issue

https://github.com/flutter/flutter/issues/121827

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/desktop_text_selection.dart
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/desktop_text_selection.dart:185]: I think that not passing this may break dark/light mode changes.  See the implementation here: https://master-api.flutter.dev/flutter/cupertino/CupertinoDesktopTextSelectionToolbarButton/CupertinoDesktopTextSelectionToolbarButton.text.html

File: packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar.dart
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar.dart:30]: Nit: I think this should be indented more like this, but not a big deal.

```dart
    CupertinoDynamicColor.withBrightness(
      color: Color(0xFFECECEC),
      darkColor: Color(0xFF333333),
    );
```

File: packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar.dart
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar.dart:77]: Nit: Can you abstract this shadow to a const variable and put it at the top with the others?  Mainly I say that just so that it's clear that these values came from the same MacOS 13.2 screenshot as the other values.

File: packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar_button.dart
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar_button.dart:50]: Can you explain why you needed to refactor this to create `child` in `build`?

File: packages/flutter/lib/src/cupertino/desktop_text_selection.dart
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/desktop_text_selection.dart:185]: I see your code below where you refactored this a bit.  It should be alright, but please try this out in both light and dark mode and make sure it works in both.

I had some light/dark mode problems with the iOS toolbar and created an example app for that (in the description of PR https://github.com/flutter/flutter/pull/117234), you could probably do something similar for the desktop Mac toolbar.

**Review by justinmc — COMMENTED** [review]
Thanks for updating this!  I hadn't noticed that it got out of date, but this version looks much better.

Some questions about light/dark mode to make sure about before merging.

Also, I wonder if there is anything that would be useful to test here?

**Comment by guidezpl** [comment]
Nice, much better. I think the text may benefit from being bumped 1-2 pts, but I don't think we'll be able to get true fidelity without using San Francisco

**Comment by luccasclezar** [comment]
@guidezpl Thanks! But yeah, I was not very sure about decreasing the size to 13, I will be bumping it to 14 again.

**Comment by luccasclezar** [comment]
@justinmc You're welcome!

Regarding the tests, I don't know. I tried to think about it when I was making these changes, but I couldn't come up with anything. Honestly, I don't know much about testing in Flutter at this scale, so what do you think?

File: packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar.dart
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar.dart:30]: @justinmc I used `dart format`, but I agree this is cleaner. Should I disregard the default formatting?

File: packages/flutter/lib/src/cupertino/desktop_text_selection.dart
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/desktop_text_selection.dart:185]: I think theme changes are OK, is this test enough or should I create the example app?

https://user-images.githubusercontent.com/12024080/224187032-94977d57-c87c-4768-9689-dad37dc37ba8.mov

**Comment by justinmc** [comment]
Ah the blur effect would be cool and I imagine would be doable in Flutter.  Maybe do it in a separate PR if that's easier, otherwise go for it here.

For tests, one thing you could do is to find the BoxDecoration and then check that it now has a BoxShadow.  You could also probably check the BorderRadius that's been added to the inner CupertinoButton.

File: packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar_button.dart
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar_button.dart:50]: For the same reason it's done on [CupertinoTextSelectionToolbarButton](https://github.com/luccasclezar/flutter/blob/2f4ca584559b33581ba771bc530879d3d2d3d725/packages/flutter/lib/src/cupertino/text_selection_toolbar_button.dart#L119), the child needs to be rebuilt when certain things change. In this case, when hovered I check if it's better for the text to be black or white depending on the background color.

**Comment by luccasclezar** [comment]
@justinmc I decided to give the blur effect a try and it's really cool.

The thing is, macOS applies a saturation boost besides the blur and without it the effect looses its charm. The only way I could find to increase saturation was via `ColorFilter.matrix`, which works, but I have to admit that I just copied and pasted the `matrixWithSat` method from here: https://github.com/flutter/flutter/issues/29483#issuecomment-477334981, instead of doing the proper thing and understanding the magic numbers.

I kept the `_matrixWithSat` method internal to `CupertinoDesktopTextSelectionToolbar` and added the above link to its documentation so that others can understand its origin.

|Just blur|Blur and saturation boost|
|-|-|
|Screenshot 2023-03-09 at 22 06 58|Screenshot 2023-03-09 at 22 16 02|

File: packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar.dart
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar.dart:77]: Don't know how I didn't think of this, I will place it with the other constants :smile:

**Comment by luccasclezar** [comment]
@justinmc And about the tests, I'm legitimately curious as to why it would be important to test "static" things like shadow and borderRadius. If someone changes these values in the future, wouldn't the change be deliberate and probably because macOS' native counterparts changed too, thus making the tests just another place to update?

**Comment by luccasclezar** [comment]
@justinmc I don't know if you didn't continue reviewing because of problems that I still need to fix, so just asking, is there something that I need to do?

I noticed that a test failed too, but I didn't bring it up as it doesn't look like there's anything to do with what I changed, the error is `failed to resolve infra/3pp/tools/cmake/linux-amd64@version:3.16.1 (line 1): no such tag`.

**Comment by gnprice** [comment]
> about the tests, I'm legitimately curious as to why it would be important to test "static" things like shadow and borderRadius. If someone changes these values in the future, wouldn't the change be deliberate and probably because macOS' native counterparts changed too, thus making the tests just another place to update?

That's one reason it could change, and it's true that in that case the test would just add another place to update. That shouldn't be a big burden, though; updating a handful of tests will be a much smaller amount of work than is involved in pinning down what the right new values are in the first place.

But the other possibility is a hypothetical future change breaking this code by accident, when the change was aimed at doing something else. For example, someone's adding some other feature, perhaps some new way to customize the appearance of this widget, and by accident they write the logic so that it never invokes some part of this code. That's the case where a test will come in handy.

I'm not sure why the you-need-a-test bot didn't comment on this PR; I would have expected it to. If it had, one thing it would have said is: the purpose of a test is to make sure we don't accidentally revert your change. Is there anything in your change where it would be a problem if it accidentally got reverted?

**Review by justinmc — COMMENTED** [review]
Sorry, I just dropped off because this got lost in my GitHub notifications.  Feel free to ping me on Discord if I disappear again :)

I'm pretty sure that test failure has nothing to do with this PR.  I tried rerunning it but I'm not sure if it will pass.  If not, can you try pushing a merge commit?

+1 to @gnprice's explanation about the tests.  I try to avoid tests that just stupidly check a value in the UI, but if there isn't a better way to test them, they're better than nothing IMO, to prevent regressions.

*Edit*: By the way, I think the blur with the contrast boost looks awesome.  Thank you for digging into that and finding the thread about how to do it.

**Comment by luccasclezar** [comment]
@gnprice Thank you for explaining all this, it makes more sense now :)

I will add some tests to make sure the widget stays like this.

**Comment by luccasclezar** [comment]
> Sorry, I just dropped off because this got lost in my GitHub notifications. Feel free to ping me on Discord if I disappear again :)

@justinmc No problem :)

> I'm pretty sure that test failure has nothing to do with this PR. I tried rerunning it but I'm not sure if it will pass. If not, can you try pushing a merge commit?

Just merged, let's see if the tests pass this time.

> +1 to @gnprice's explanation about the tests. I try to avoid tests that just stupidly check a value in the UI, but if there isn't a better way to test them, they're better than nothing IMO, to prevent regressions.

Yeah, it makes a lot of sense :)

> Edit: By the way, I think the blur with the contrast boost looks awesome. Thank you for digging into that and finding the thread about how to do it.

Thank you! It took some time to understand why a simple blur looked wrong and how to fix it, but I think the result was pretty good too.

File: packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar.dart
Comment (luccasclezar) [reply packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar.dart:81]: The empty comment prevents the formatter from placing each number in a separate line. Would it be better if I just didn't use the formatter anymore?

**Comment by luccasclezar** [comment]
There are two tests failing on web:

1. `ImageFilter.compose not implemented for CanvasKit`
2. `ImageFilter.erode not implemented for HTML renderer`

Canvaskit does not support `ImageFilter.compose` yet (https://github.com/flutter/flutter/issues/120123) and HTML renderer does not support `ImageFilter.erode` (which I don't use explicitly but it's probably used by another ImageFilter).

The only solution that I think of is to use only blur on web.

File: packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar.dart
Comment (justinmc) [reply packages/flutter/lib/src/cupertino/desktop_text_selection_toolbar.dart]: Ah, I think it's fine like this 👍

**Comment by justinmc** [comment]
I agree.  Can you disable the unsupported pieces on web and leave a comment explaining and linking to that GitHub issue?

Actually this won't affect many users.  By default, Flutter web uses the browser's context menu.  I only added support for showing this Flutter-drawn context menu on web in the last few months.

**Comment by justinmc** [comment]
I still see the CanvasKit failure happening: `ImageFilter.compose not implemented for CanvasKit.`

**Comment by luccasclezar** [comment]
@justinmc Yeah, I was trying to figure it out but I don't have a clue about why it's still failing 😅

As you can see [here](https://github.com/flutter/flutter/pull/121829/files#diff-5c2e069a8b9bbd50e68038089a5d3160f2d130fb1df0a2904a9c2ebeecc8d61bR109-R122), I am checking for kIsWeb and not using ImageFilter.compose if it is.

**Comment by justinmc** [comment]
Is it this call to `.compose` in the test itself? https://github.com/flutter/flutter/pull/121829/files#diff-dce0386c60da0c7fcd6c25a72255769ee9d4ef9550f1a7aa283b86b37f3cbaa5R43

**Comment by luccasclezar** [comment]
@justinmc Thank you for pointing that out 😅

Just fixed it, let's see how it goes.

File: packages/flutter/test/cupertino/desktop_text_selection_toolbar_test.dart
Comment (justinmc) [inline packages/flutter/test/cupertino/desktop_text_selection_toolbar_test.dart:42]: Nit: "outer" => "Outer"

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Thanks for all of your work on this!  It's going to look awesome.  I'll try to get a second reviewer so we can merge this.

**Comment by luccasclezar** [comment]
You're welcome! And thank you for your suggestions and advices!

File: packages/flutter/test/cupertino/desktop_text_selection_toolbar_test.dart
Comment (MitchellGoodwin) [inline packages/flutter/test/cupertino/desktop_text_selection_toolbar_test.dart:40]: Nit: "it's" => "its"

**Review by MitchellGoodwin — APPROVED** [review]
LGTM. The new menu looks great!

File: packages/flutter/test/cupertino/desktop_text_selection_toolbar_test.dart
Comment (luccasclezar) [reply packages/flutter/test/cupertino/desktop_text_selection_toolbar_test.dart]: Fixed

File: packages/flutter/test/cupertino/desktop_text_selection_toolbar_test.dart
Comment (luccasclezar) [reply packages/flutter/test/cupertino/desktop_text_selection_toolbar_test.dart]: Fixed

**Comment by luccasclezar** [comment]
@MitchellGoodwin Thank you! 😁

**Comment by justinmc** [comment]
@luccasclezar Am I not getting the blur effect for some reason or is it just subtle here?

Screenshot 2023-04-21 at 4 35 35 PM

**Comment by tgucio** [comment]
@justinmc it does appear to blur but I guess it's subtle with the blur amount applied:

toolbar

Thanks for the PR @luccasclezar!

**Comment by luccasclezar** [comment]
@justinmc @tgucio Exactly. The macOS native blur effect is almost invisible when the colored areas on the background are small, but when they are larger it's very noticeable.

**Comment by justinmc** [comment]
Ah thanks, you guys are right!  I tried it on Mac and Flutter and they're both similarly subtle. 👍

**Pull request merged**
